Add query functions to function_score

brwe opened the pull request:
Allows to combine query scores with mult, sum etc
by wrapping individual queries in a `query_function`
of a `function_score like` so:

[Edit, see discussion below]

```
{
  "query": {
    "function_score": {
      "score_mode": "multiply",
      "functions": [
        {
          "query": {
            "match": {
              "text": "cat"
            }
          }
        },
        {
          "query": {
            "match": {
              "text": "dog"
            }
          }
        }
      ],
      "boost_mode": "replace"
    }
  }
}
```

[Was initially: ]

```
{
  "query": {
    "function_score": {
      "score_mode": "multiply",
      "functions": [
        {
          "query_function": {
            "query": {
              "match": {
                "text": "cat"
 ...
```

relates to #17116

brwe commented:
I am a little unsure about this implementation and added three TODOs where I am particularly unsure.

JnBrymn-EB commented:
Per https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-function-score-query.html, the current `function_score` syntax is 
```
        {
            "filter": {},
            "FUNCTION": {}, 
            "weight": number
        }
```
Rather than adding a new type of function, would it be more consistent to add a parameter at this level called `query` (similar to the existing `filter`). This would require one less level of nesting if the function was omitted.

brwe commented:
So you mean the whole thing would more look like this? :

```
{
  "query": {
    "function_score": {
      "score_mode": "multiply",
      "functions": [
        {
          "filter": {
            "match": {
              "some_other_field": "foo"
            }
          }, 
          "query": {
            "match": {
              "text": "cat"
            }
          }, 
          "weight": 2
        },
       ...
      ],
      "boost_mode": "replace"
    }
  }
}

```
That might indeed be easier to read.

Or do you mean make the query a general option for all functions, like:

```
{
  "query": {
    "function_score": {
      "score_mode": "multiply",
      "functions": [
        {
          "filter": {
            "match": {
              "some_other_field": "foo"
            }
          },
          "field_value_factor": {
            "field": "num_field"
          },
          "query": {
            "match": {
              "text": "dog"
            }
          },
          "weight": 2
        }
        ...
      ],
      "boost_mode": "replace"
    }
  }
}
```
That was the original idea in #17116 but we discarded it because then we would need to know how to combine query score with the other function (in this case the output of `field_value_factor`).

JnBrymn-EB commented:
I had in mind something that looks like your first example, but I was also wondering if it was sensible to combine this with the existing function types - probably doesn't. So yes, the first option. Effectively I think you're implementing this as a function_score function type called `query`, yes?

So far I'm ok with all of the above.

brwe commented:
Added a new commit to change syntax accordingly (f9876fd).

clintongormley commented:
@brwe sorry for the late notice but I'm not sure I agree with the latest changed.  I don't understand what output you'd expect from that construct.  With the original form (with `query` as a type of function) you understand that the result for each doc is its score.

I think this only really becomes useful once we have the second part (combining scores from each function with a script)

clintongormley commented:
>  I'm not sure I agree with the latest changed.

Sorry I misread the history. I DO agree with the current format :)  The only thing I'd do is to change the name from `query` to `query_score`.  We already have `filter`, so `query` sounds too much like it can be used like a filter, while `query_score` is closer to `script_score`.

brwe commented:
OK, changed the syntax to be:


```
{
  "query": {
    "function_score": {
      "score_mode": "multiply",
      "functions": [
        {
          "filter": {
            "match": {
              "some_other_field": "foo"
            }
          }, 
          "query_score": {
            "match": {
              "text": "cat"
            }
          }, 
          "weight": 2
        },
       ...
      ],
      "boost_mode": "replace"
    }
  }
}

```

I need help with the TODOs. @jpountz when you have a minute, could you take a look? If not let me know and I will find someone else.

jpountz commented on core/src/main/java/org/elasticsearch/common/lucene/search/function/QueryFunction.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    @Override
+    public void initWeight(IndexSearcher searcher, boolean needsScores) throws IOException {
```
you can remove the needsScores parameter, it should always be true?

jpountz replied:
maybe we should remove this method entirely and provide the IndexSearcher and create the weight in the constructor?

brwe replied:
We do not have the IndexSearcher when we create the ScoreFunctions (in FunctionScoreQueryBuilder.doToQuery). Do you mean delay the constructor until we call weight() and remember the QueryShardContext until then? Or can I get an IndexSearcher from somewhere else?

jpountz commented on core/src/main/java/org/elasticsearch/common/lucene/search/function/QueryFunction.java:
```diff
@@ -0,0 +1,101 @@
+
+    @Override
+    public void initWeight(IndexSearcher searcher, boolean needsScores) throws IOException {
+        // TODO: which weight should be created here? normalized or not?
```
normalized

jpountz commented on core/src/main/java/org/elasticsearch/common/lucene/search/function/QueryFunction.java:
```diff
@@ -0,0 +1,101 @@
+    @Override
+    public LeafScoreFunction getLeafScoreFunction(LeafReaderContext ctx) throws IOException {
+        final Scorer scorer = weight.scorer(ctx);
+        // TODO: not sure if this is the most efficient way to do it
```
looks great :)

jpountz commented on core/src/main/java/org/elasticsearch/common/lucene/search/function/QueryFunction.java:
```diff
@@ -0,0 +1,101 @@
+        QueryFunction that = (QueryFunction) o;
+        if (needScores != that.needScores) return false;
+        if (query != null ? !query.equals(that.query) : that.query != null) return false;
+        return weight != null ? weight.equals(that.weight) : that.weight == null;
```
weights do not implement equals/hashcode so I think we should only compare the query

jpountz replied:
actually we should probably use weight.getQuery() for equality and hashCode. This will be the rewritten query.

brwe replied:
ok, it now uses query and weight.query

jpountz commented on core/src/main/java/org/elasticsearch/common/lucene/search/function/FiltersFunctionScoreQuery.java:
```diff
@@ -144,6 +144,9 @@ public Weight createWeight(IndexSearcher searcher, boolean needsScores) throws I
             subQueryNeedsScores |= filterFunctions[i].function.needsScores();
             filterWeights[i] = searcher.createNormalizedWeight(filterFunctions[i].filter, false);
+            //TODO: Not sure if we should add such a method because it does nothing for any score function except for query_function but
+            // I don't have a good idea where else to put it
```
I left a comment in QueryFunction: I think we can get rid of this method by passing the IndexSearcher in the constructor

jpountz commented on docs/reference/query-dsl/function-score-query.asciidoc:
```diff
@@ -415,6 +415,17 @@ Example:
+"query_score": {
+    ARBITRARY QUERY
+}
+--------------------------------------------------
```
I think it is important to mention that this produces a score of 0 if the query does not match.

brwe replied:
ok, I added a note

jpountz commented:
@brwe I left some comments!

brwe commented:
@jpountz thanks for the review! Addressed all comments except for the IndexSearcher in constructor of QueryFunction because I am not sure where to get it from, see https://github.com/elastic/elasticsearch/pull/18612#discussion_r65726645 . Will try to find a different way on Monday.

mckinnovations commented:
Any timeline for this feature ? when is it going to be released?

dakrone commented:
@jpountz I think this needs a re-review?

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

dakrone commented:
@jpountz you were reviewing this, is this something we should bring up to date?

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

dakrone commented:
Closing this as @brwe isn't around to merge it, feel free to re-open if we want to revisit this.
